gamification/models.py

```diff
@@ -0,0 +1,182 @@
+
+        :param points:          an integer value representing the
+                                points earned by the participant
+                                for performing an activity.
+        :param activity_string: a string representing the activity
+                                performed by the participant.
+        :param performed_at:    a datetime object representing the date
+                                and time when this activity was performed.
+        :param updated_at:      a datetime object representing the date and time when
+                                this activity was performed.
+        """
+        self.update_score_and_level(points)
+        self.add_activity(points,
+                          activity_string,
+                          performed_at,
+                          updated_at,
+                          )
+
+    def deduct_points(self, points_to_deduct, activity_string,
+                      performed_at, updated_at):
+        """
+        Deduct points for performing some specific activities.
+        """
+        self.add_points(points_to_deduct,
+                        activity_string,
+                        performed_at,
+                        updated_at,
+                        )
+
+    def find_level_for_score(self, score):
+        """
+        Find suitable level based on the total score earned.
+        """
+        level = Level.objects.get(min_score__lte=score, max_score__gt=score)
+        return level
+
+    def update_score_and_level(self, points):
+        """
+        Update score and level based on points.
+        """
+        if points < 0 and self.score < abs(points):
+            self.score = 0
+        else:
+            self.score += points
+
+        self.level = self.find_level_for_score(self.score)
+        self.save()
+
+    def add_activity(self, points, activity_string, performed_at, updated_at):
+        """
+        Add a new activity to the participant.
+        """
+        activity = Activity.objects.create(
+            name=activity_string,
+            points=points,
+            performer=self,
+            performed_at=performed_at,
+            updated_at=updated_at)
+        self.activities.add(activity)
+
+    def find_badges_for_activity(self, activities):
+        """
+        Find the badge based on the activities peformed by the participant.
+
+        :param activities: a QuerySet dict containing the 'name'
+                           as key and 'name of the activity' as value.
+        :return:           a badge object.
+        """
+        activities = [activity['name'] for activity in activities]
+        badge_objects_list = []
+        badges = Badge.objects.all()
+        for badge in badges:
+            b_activities = badge.b_activities.values('name')
+            b_activities = [b_activity['name'] for b_activity in b_activities]
+            # Skip the badge which have no activities
+            if not len(b_activities):
```

Comment:
which badges have no activities?

I guess this should at least be a 'info' level log entry